### PR TITLE
Commit hacky script to visualise store inheritance

### PR DESCRIPTION
**Pull request by DMRobertson**
Use e.g. with `scripts-dev/storage_inheritance.py DataStore --show`.

It's dirty, but it's useful enough to help disentangle the sphagetti.

### CODE REVIEW

**Comment by DMRobertson** [comment]
`scripts-dev/storage_inheritance.py DataStore --show` running at 9e361c8:

image

`scripts-dev/storage_inheritance.py BaseSlavedStore --show` at cd7257bf27b0344266ae507085066c9c6c604b6d:

image

`scripts-dev/storage_inheritance.py --show` at 9e361c8:

image

File: scripts-dev/storage_inheritance.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+#! /usr/bin/env python3
+import argparse
+import os
+import re
+import subprocess
+import sys
+import tempfile
+from typing import Iterable, Optional, Set
+
+import networkx
```
Comment (babolivier) [inline scripts-dev/storage_inheritance.py:10]: We should add this to the dev dependencies.

File: scripts-dev/storage_inheritance.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            "rg",
+            "-o",
+            "--no-line-number",
+            "--no-filename",
+            "--multiline",
+            r"class .*Store\((.|\n)*?\):$",
+            "synapse",
+            "tests",
+        ],
+        cwd="/home/dmr/workspace/synapse/",
```
Comment (babolivier) [inline scripts-dev/storage_inheritance.py:35]: I assume we probably want to change this so it doesn't point to your home directory.

**Comment by babolivier** [comment]
Also I would suggest writing the file on disk rather than opening it. For example I know Gnome has some weird magic around default apps, and when I tried running the script it opened it with Gimp for some reason despite the fact that I've changed the default app for PNGs to something else some time ago.

**Comment by DMRobertson** [comment]
> Also I would suggest writing the file on disk rather than opening it. 

There's a command line option for this: `--output DEST`.

> For example I know Gnome has some weird magic around default apps, and when I tried running the script it opened it with Gimp for some reason despite the fact that I've changed the default app for PNGs to something else some time ago.

Suggest checking your config with:

```
$ xdg-mime query default image/png
org.gnome.eog.desktop
```

**Comment by babolivier** [comment]
> There's a command line option for this: `--output DEST`.

Hah, I see. Maybe worth adding some doc along with the script so we remember how to use it, then? Otherwise I'm afraid it'll become like the config validator/parser util in Synapse and we'll completely forget it's a thing until someone complains we don't have such a tool or it's broken.

**Comment by DMRobertson** [comment]
> Hah, I see. Maybe worth adding some doc along with the script so we remember how to use it, then?

I use argparse so there's a `--help` out of the box for this

**Comment by anoadragon453** [comment]
What's the expected lifetime of this script? Do we expect to remove it once DataStore is less :spaghetti:?

**Comment by clokep** [comment]
I'm going to remove this from the review queue since it seems to be awaiting info from @DMRobertson.

**Comment by DMRobertson** [comment]
> What's the expected lifetime of this script? Do we expect to remove it once DataStore is less spaghetti?

I hadn't really considered it, but that sounds sensible.

(I can imagine a version that imports a class and inspects `__mro__` attributes to build this graph, but hopefully we aren't relying on inheritance to this extent anywhere else?)

**Pull request closed without merging**